Disallow auto_element_wise for in-place and fix some in-place gradients

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79056
* __->__ #85634

Fixes https://github.com/pytorch/pytorch/issues/85535

Also fixes the backward and forward gradients of `nn.functional.threshold`. The issue was that in-place gradients weren't tested because the in-place variants were not properly registered to the OpInfo.

Perhaps an alternative to this to make auto_element_wise smart enough to actually handle the in-places cases (we have 4 cases total now where we manually copy_ after doing auto_element_wise), but that requires a few more changes.

soulitzer committed:
Disallow auto_element_wise for in-place

[ghstack-poisoned]

soulitzer committed:
Update on "Disallow auto_element_wise for in-place"


Fixes https://github.com/pytorch/pytorch/issues/85535

[ghstack-poisoned]

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -2020,8 +2020,8 @@
 - name: threshold_(Tensor(a!) self, Scalar threshold, Scalar value) -> Tensor(a!)
-  self: threshold_backward(grad, result, threshold)
-  result: auto_element_wise
+  self: threshold_backward(grad, self, threshold)
```
That's not ok I think.
If you do `threshold(self, a, b)` with `b > a`, this will be wrong!

soulitzer replied:
Oh I think that was happening before this change, but not anymore? 
We want to do `grad_input = self > threshold ? grad_output : zero` right?

albanD replied:
1. We should have such a case in the OpInfo samples
2. We want to check self yes IIRC (use OpInfo testing to see which one is correct to be sure)
3. Changing this can have significant impact on memory usage patterns. This is the inplace relu op that is used everywhere. Using `self` here means that it will be cloned: extra buffer!

soulitzer replied:
> 1. We should have such a case in the OpInfo samples
> 2. We want to check self yes IIRC (use OpInfo testing to see which one is correct to be sure)

We don't have an explicit check because all values are chosen at random, but that means it will be checked with high probability.

> 3. Changing this can have significant impact on memory usage patterns. This is the inplace relu op that is used everywhere. Using self here means that it will be cloned: extra buffer!

Okay I checked and relu doesn't actually use threshold (and since a == b for relu, so we can still be better off saving result instead of self). For threshold though I guess that now it makes less sense to use in-place (at least when gradients are required).

albanD reviewed (commented):
The change to the backward formula sounds wrong.

soulitzer committed:
Update on "Disable torch.library.Library with PYTORCH_DISABLE_LIBRARY"




[ghstack-poisoned]

soulitzer committed:
Update on "Disallow auto_element_wise for in-place and fix some in-place gradients"


Fixes https://github.com/pytorch/pytorch/issues/85535

Also fixes the backward and forward gradients of `nn.functional.threshold`. The issue was that in-place gradients weren't tested because the in-place variants were not properly registered to the OpInfo.

Perhaps an alternative to this to make auto_element_wise smart enough to actually handle the in-places cases (we have 4 cases total now where we manually copy_ after doing auto_element_wise), but that requires a few more changes.

[ghstack-poisoned]

albanD reviewed (approved):
SGTM

soulitzer commented:
@pytorchbot merge -g
